db/db_impl.cc

```diff
@@ -1011,6 +1011,20 @@ void DBImpl::PurgeObsoleteFiles(const JobContext& state, bool schedule_only) {
       case kLogFile:
         keep = ((number >= state.log_number) ||
                 (number == state.prev_log_number));
+
+        /* We need to check if the log file is already added to recycle list
+         * before we mark it for deletion.
+         */
+        if (!keep && (std::find(log_recycle_files.begin(),
+            log_recycle_files.end(),number) != log_recycle_files.end())) {
```

Comment:
@siying Entire PurgeObsoleteFiles is not under any lock , it will be racy but probably not harmful as deletion is happening only from here. Already other vector like log_delete_files , full_scan_candidate_files etc. are accessed from this function without locking.